pandas/tslib.pyx

```diff
@@ -4386,6 +4387,40 @@ cpdef normalize_date(object dt):
         raise TypeError('Unrecognized type: %s' % type(dt))
 
 
+@cython.wraparound(False)
+def shift_months(int64_t[:] dtindex, int months):
+    '''
+    Given an int64-based datetime index, shift all elments
+    specified number of months using relativedelta semantics
+    '''
+    cdef:
+        Py_ssize_t i
+        int isleap, days_in_month, new_month
+        pandas_datetimestruct dts
+        int count = len(dtindex)
+        int64_t[:] out = np.empty(count, dtype='int64')
+
+    count = len(dtindex)
+    out = np.empty(count, dtype='int64')
+
+    for i in range(count):
+        if dtindex[i] == NPY_NAT:
+            out[i] = NPY_NAT
+        else:
+            pandas_datetime_to_datetimestruct(dtindex[i], PANDAS_FR_ns, &dts)
```

Comment:
this routine could actually be used in the Dateoffset for adding months (possibly dealing with month end/start as well)

could do in another PR though